lib/ClangImporter/ImportType.cpp

```diff
@@ -1771,15 +1771,29 @@ adjustResultTypeForThrowingFunction(ForeignErrorConvention::Info errorInfo,
   switch (errorInfo.TheKind) {
   case ForeignErrorConvention::ZeroResult:
   case ForeignErrorConvention::NonZeroResult:
+    // Check for a bad override.
+    if (resultTy->isVoid())
+      return Type();
     return TupleType::getEmpty(resultTy->getASTContext());
 
   case ForeignErrorConvention::NilResult:
-    resultTy = resultTy->getAnyOptionalObjectType();
-    assert(resultTy &&
-           "result type of NilResult convention was not imported as optional");
-    return resultTy;
+    if (Type unwrappedTy = resultTy->getAnyOptionalObjectType())
+      return unwrappedTy;
+    // It's possible an Objective-C method overrides the base method to never
+    // fail, and marks the method _Nonnull to indicate that. Swift can't
+    // represent that, but it shouldn't fall over either.
+    if (resultTy->isAnyClassReferenceType()
+        || resultTy->is<AnyFunctionType>()
+        || resultTy->getAnyPointerElementType())
+      return resultTy;
+    return Type();
 
   case ForeignErrorConvention::ZeroPreservedResult:
+    // Check for a bad override.
+    if (resultTy->isVoid())
+      return Type();
+    return resultTy;
```

Comment:
does this mean that if a non-nil return type is returned the error is checked first in the generated code for the NSError?

e.g.

```objective-c
- (nonnull id)foo:(NSError **_Nullable)error;
...

NSError *error;
id result = [something foo:&error];
if (error) { 
    assumeItFailed(); 
} else {
    useResult(result);
} 
```

If so that might cause unexpected failures of un-initialized memory cases when passing into existing failure methods.

Or is this just for the type checker?